### DISCUSSION TITLE
Does Fluvio have any interoperability with Kafka?

### DISCUSSION THREAD

**Discussion by frol** [fluvio-community/fluvio#1195] [Q&A] [answered]
I have found a few references about Kafka protocol in the source code, but I don't believe those few places would be enough to ensure compatibility.

P.S. I am excited to see Fluvio project (Rust and Wasm is the perfect match!)

**Comment by sehz** [comment, marked as answer]
Hi, Welcome to Fluvio Project.  

We used to have strict compatibility with Kafka but decided to abandon it due to maintenance costs.  In addition, it prevented us from adding exciting new capabilities such as SmartStream.

Currently, we still use a subset of Kafka protocol with a compatible storage format but no longer consider compatible with Kafka.   We created separate repo where we still maintain Kafka compatible protocol Rust: https://github.com/infinyon/flv-kf-protocol.  If there is demand, we may consider creating a Kafka connector, but only if there is sufficient demand.  

We are looking forward to feedback and contribution to projects.

**Reply by frol** [reply]
Working on [NEAR protocol](https://near.org), I am considering a way to expose real-time events happening in the network to the application developers. Kafka is the main contender at the moment, but Fluvio might be a great option as well, though at the current stage we are looking for something that is mature. I will look forward to seeing how Fluvio evolves! Having Wasm processors opens up many other opportunities for application developers.

**Reply by sehz** [reply]
We had some interest from other blockchain developers as well.  Are there use cases or features that might worthwhile for projects like yours?

**Reply by frol** [reply]
The blockchain main operation is to maintain the state and arrive at a consensus. Applications need that latest state, but they often also need to react to certain events and either maintain some computed aggregated value or just notify end-users. The blockchain applications are "serverless", and thus there are quite common scenarios when the frontend is a static single-page application, and thus developers don't even maintain servers. Thus, we are looking for solutions that could efficiently store, filter, and stream events to web-browsers (e.g. websocket).

**Reply by sehz** [reply]
This certainly fits Fluvio's key use cases.    In this case, Fluvio can consume events from NEAR and put into topic streams where user can query or compute "aggregated" values using SmartStream.  We are working on adding [aggegration](https://github.com/infinyon/fluvio/issues/1173) features.   

In addition, we have experimental [WebSocket Client](https://github.com/infinyon/fluvio-client-wasm), which is packaged as WASM modules in the browser.  It needs some polishing, but it is usable now.   Users can directly consume/publish events from Fluvio (including SmartStream) through WebSocket with the WASM client.   We are using it for the next version of cloud UI, which will be coming out shortly. 

If you are interested,  maybe we can collaborate on building a NEAR connector (for pulling NEAR events) for Fluvio as the first step?

**Reply by gadLinux** [reply]
We will not even test Fluvio without proper Kafka support. We work with Confluent's excellent Kafka Streams API and we don't see we can opt-out this tech once there. Since there are no alternatives. Fluvio seemed like a great opportunity but having no support for Kafka is a no go. I suggest developing and enhancing Kafka support to reach at least to the same level of operability that exists with Kafka Streams...  Is there any chance to get it?

**Reply by sehz** [reply]
Hi, is this about Fluvio connecting to Kafka (in order to consume/produce) or providing Kafka compatible API on top of Fluvio in order to make it easy to migrate apps?

**Reply by gadLinux** [reply]
Is more about be bale to consume and produce to Kafka. We use it for streaming and there's a lot of work done processing Kafka Streams. In our side it would be great to be able test and compare Fluvio against Kafka Streams. 
I don't know if others had other requirements. For us API compatibility is desiderable but not required.

**Reply by sehz** [reply]
Gotcha.  We are looking to add add kafka connector to roadmap

**Reply by saminahbab** [reply]
Yeah, personally, I see this as quite important myself too. 
Kafka is the dominant pub/sub tool in the industry. I feel its unreasonable not to have first class support for kafka. 
Am I right in saying that there are no ways at the moment to, at least, use kafka as some kind of streaming source?

**Reply by sehz** [reply]
Thanks for the feedback.  We actually had some users who had successfully able to connect Kafka with Fluvio.  We are looking to add that as a supported connector.

**Comment by ajhunyady** [comment]
@frol, Kafka sink & source connectors are now ready for use: https://github.com/infinyon/fluvio-connectors/pull/256.
The documentation is in the connector section: https://www.fluvio.io/connectors/

**Reply by frol** [reply]
@ajhunyady Thanks for letting me know! We have already redesigned the system, but I will keep fluvio in mind for my next projects!
